src/riot-rs-embassy-common/src/identity.rs

```diff
@@ -0,0 +1,94 @@
+//! General structure of a device identities.
+
+/// Describes how a board produces unique identifiers.
+pub trait DeviceId {
+    /// Opaque type representing a unique-ish identifier of a board.
+    ///
+    /// See [`.get()`](Self::get) for precise semantics.
+    ///
+    /// # Open questions
+    ///
+    /// Do we have any concrete serializeers we want to prescribe? Or should we switch to a const
+    /// size ([u8; N]) anyway?
+    type DeviceId: core::fmt::Debug + defmt::Format;
+
+    /// Error type indicating that no identifier is available.
+    ///
+    /// This is encouraged to be [`core::convert::Infallible`] where possible.
+    ///
+    /// # Open questions
+    ///
+    /// Some architectures will have to read this (eg. at QSPI initialization time); is there guidance
+    /// on how to report "Not yet available"?
+    type Error: core::error::Error + defmt::Format;
+
+    /// Obtains a unique identifier of the device.
+    ///
+    /// A successful return value is, within the scope of the active board, reasonably unique. (The
+    /// "reasonably" part is what allows this to be used with devices whose vendors do not assign
+    /// serial numbers but large (>= 64 bits) random identifiers).
+    ///
+    /// This value may be obtained from the processor itself, or from a peripheral that is connected to
+    /// the processor as part of the board.
+    ///
+    /// For the type implementing this trait at its conventional position
+    /// `riot_rs::arch::identity::DeviceId`, a convenience function to call it exists at
+    /// `riot_rs::identity::device_identity()`.
+    ///
+    /// # Open questions
+    ///
+    /// Uniqueness is currently described in the scope of the board, but this is implemented as part of
+    /// the `arch`. Implementation experience will show what is more realistic.
+    ///
+    /// Ideally, those should be coordinated even outside of the project, so that a hypothetical
+    /// extension to probe-rs that connects to devices by their ID can reuse the identifiers. This will
+    /// require some more agreement on their structure as well as their scope. (probe-rs does not work
+    /// on boards but just on chips).
+    ///
+    /// # Errors
+    ///
+    /// This prodcues an error if no device ID is available on this board, or is not implemented.
+    fn get() -> Result<Self::DeviceId, Self::Error>;
+}
+
+/// A type implementing [`DeviceId`] that always errs.
+///
+/// This can be used both on architectures that do not have a unique identifier on their boards,
+/// and when it has not yet been implemented.
+pub struct NoDeviceId<E: core::error::Error + defmt::Format + Default>(
+    core::marker::PhantomData<E>,
```

Comment:
There are two ways of using this, `NoDeviceId<NotImplemented>` and `NoDeviceId<NotAvailable>`. The PhantomData is there because if it were just a `pub struct NoDeviceId<E: …>;`, Rust would complain that this is a needless generic, and the Phantom justifies it. (It's not like instances of this are relevant anyway).

The only alternative I see to a PhantomData here is to have two structs NoDeviceIdNotImplemented and NoDeviceIdNotAvailable, which boils down to probably the same binary but I think it's prettier being generic (like, "it never works because of E").

Reply:
For the two (core::error::Error + defmt::Format) needed for the DeviceId::Error type (or the three for NoDeviceId) I think creating an extra trait would be excessive, not only because it's just replacing few occurrences with more code, but also because it'd mean that the extra trait would need to be explicitly implemented on core::convert::Infallible, and all who'd use some other preexisting type that already implements core::convert::Error and defmt::format (which is not too uncommon for embedded errors) would need to empty impl that new trait too.